kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+            for (proc_index, value) in procs.iter().enumerate() {
+                if proc_index < new_process_count {
+                    let process_start_address = value.unwrap().get_addresses().flash_start;
+                    let process_end_address = value.unwrap().get_addresses().flash_end;
+
+                    if new_process_end_address >= process_start_address
+                        && new_process_end_address <= process_end_address
+                    {
+                        /* Case 1
+                         *              _________________          _______________           _________________
+                         *  ___________|__               |        |              _|_________|__               |
+                         * |           |  |              |        |             | |         |  |              |
+                         * |   new app |  |  app2        |   or   |   app1      | | new app |  |  app2        |
+                         * |___________|__|              |        |             |_|_________|__|              |
+                         *             |_________________|        |_______________|         |_________________|
+                         *
+                         * ^...........^                                           ^........^
+                         * In this case, we discard this region and try to find another start address from the end address + 1 of app2
+                         */
+                        return Err((process_end_address + 1, ProcessLoadError::NotEnoughMemory));
+                    } else if new_process_start_address >= process_start_address
+                        && new_process_start_address <= process_end_address
+                    {
+                        /* Case 2
+                         *              _________________
+                         *  ___________|__               |    _______________
+                         * |           |  |              |   |               |
+                         * |   app2    |  |  new app     |   |     app3      |
+                         * |___________|__|              |   |_______________|
+                         *             |_________________|
+                         *
+                         *                 ^
+                         *                 | In this case, the start address of new app is replaced by 'the end address + 1' of app2,
+                         *                   and we try to find another start address from the end address + 1 of app2 and recheck for
+                         *                   the previous condition
+                         */
+                        return Err((process_end_address + 1, ProcessLoadError::NotEnoughMemory));
+                    }
+                }
+            }
+            Ok(())
+        });
+        Ok(())
+    }
+
+    fn find_next_available_address(&self, new_app: &mut NewApp) -> Result<(), ErrorCode> {
```

Comment:
I think this function should take a slice and return the values, not take in a mutable object.

Reply:
I think I've implemented it, please check.